docs/design/Eras.md

```diff
@@ -0,0 +1,56 @@
+| 2020/06/05 | [Hard Fork Combinator II: Time for a Hard Fork](https://drive.google.com/file/d/1QIJ-VBlj-txB6K6E7DIEnY5TzaD89qQm/view) | As any computer scientist, physicist, or cross-timezone remote worker can testify, time is a tricky and unforgiving beast. Cardano, as a slot-based Proof of Stake system, is inherently time-driven, so it is vital to get time right. Across the hardfork from Byron to Shelley, the way in which time is divided into slots will change, which turned out to be a massive complicating factor. In this presentation, Edsko will show us how the hard fork combinator handles time across the fork point. |
+
+In a nutshell, pretty much all types in `Cardano.Api` are indexed by the `Era`. The node will provide a `TimeInterpreter` object by the `LocalStateQuery` protocol, which allows `cardano-wallet` to calculate what time a certain slot occurs at. Calculating the time of a given slot number is more complicated than it sounds.
+
+## Configuration
+
+Both `cardano-node` and `cardano-wallet` have command-line parameters and/or configuration options for the hard-fork combinator.
+
+### Genesis
+
+Nodes must be configured with genesis blocks. The genesis block of Cardano is the Byron genesis block. Some subsequent eras (Shelley and Alonzo, at the time of writing) also have their own genesis block.
+
+Only the header hash of a genesis block is stored on-chain.
+
+Note that (annoyingly) there are two blocks with epoch 0 and slot 0. One of these is the genesis block, and the other is the first real block, whose parent header hash is the genesis block header hash.
+
+When `cardano-wallet` connects to its local `cardano-node`, it must provide a header hash for the byron genesis block. Header hashes for subsequent eras are contained on-chain in the voting system update proposals.
+
+For _mainnet_, we have hardcoded the genesis block hash. It will never change. For _staging_ and _testnet_, users must supply a genesis block to the [[cli]], so that its hash may be used when connecting to the local node.
+
+### _Instafork_™ Technology
+
+Hard forking to Alonzo (the latest era, at the time of writing) by natural methods requires several epochs' worth of network operation time to take effect. For disposable local testnets, this would delay startup and make our integration tests slower. Therefore, `cardano-node` can be configured to hard fork itself at the start of any given epoch number, without an update proposal. It's also possible to hard fork directly into all the eras at epoch 0. This is what we use for the integration tests `local-cluster`.
```

Comment:
```suggestion
Hard forking to Alonzo (the latest era, at the time of writing) by natural methods requires several epochs' worth of network operation time to take effect. For disposable local testnets, this would delay startup and make our integration tests slower. Therefore, `cardano-node` can be configured to hard fork itself at the start of any given epoch number, without an update proposal. It's also possible to hard fork directly into all the eras at epoch 0. This is what we use for the integration tests `local-cluster`. (The configuration options are named `TestShelleyHardForkAtEpoch` and similar.)
```

Reply:
Yeah, I'm not quite sure of the exact story, but roadmap eras are quite broad. Goguen Era needed to be split into multiple hard forks. In the source code however, a type with an `IsCardanoEra` instance represents a hard fork, e.g. [Cardano.Api.AlonzoEra](https://input-output-hk.github.io/cardano-node/cardano-api/lib/Cardano-Api.html#t:AlonzoEra). So a roadmap era is different to a `Cardano.Api` era. We shall need to ask/dig around to clarify this.

Reply:
> So a roadmap era is different to a Cardano.Api era

I also believe that it's unfortunate overloading of the word "era" that has occurred.